### ISSUE TITLE
Sound issue with 1.2.7

### ISSUE THREAD

**Issue by JasonMcRay** [rwtema/extrautilities_old_issues#671]
This one is bit wierd but I pinpointed it at Extra Utilities after some testing. The issue is, whenever I open my inventory I hear in my case IC2/GT Hammer and IC2/GT Drill sound. I know this is strange. Testiing was done with just IC2, GT, NEI and ExU with Forge 1420

**Comment by Leftorius** [comment]
can confirm this :I

**Comment by SuperSandro2000** [comment]
Can confirm this, too!

**Comment by Peach774** [comment]
IS there some way to fix this? Also, which mod is causing it, because I know someone who is having the same problem without IC2 or Gregtech. Could it be NEI?

**Comment by SuperSandro2000** [comment]
Downgrade to 1.2.5.

**Comment by rxiv** [comment]
Issue exists with 1.2.8 also.

**Comment by SuperSandro2000** [comment]
I used Forge 1448 and sometimes i heard the sound even if I only moved my mouse.

**Comment by Blood-Asp** [comment]
Maybe this helps searching for the problem:
http://forum.industrial-craft.net/index.php?page=Thread&postID=190418#post190418

**Comment by bentalexanderhaase** [comment]
Wow finally found it! This bug was grinding our nerves to the extremes. This could also be related to these kind of crashes: http://openeye.openmods.info/crashes/c2efef2765162477f951507848365d72

**Comment by Dream-Master** [comment]
I think the hammering sound is from the Minecraft Anvil but i am not 100 % sure.

**Comment by Dream-Master** [comment]
Still exists with Version 1.2.10

**Comment by leagris** [comment]
Still exists with Version 1.2.11, now even cause Concurrent Modification Exception and crash client if player inventory is opened while sound plays.

See: crash report http://pastebin.com/H0r5Ni6i

**Comment by Dream-Master** [comment]
Maybe we have to contact RWThema via Twitter because he never answered here.

**Comment by KageDragon** [comment]
I also have this issue, with 1.2.7 or higher, and eventually the ConcurrentModificationException crash.

I found how to get Minecraft to show me some more info on sounds, and found this, whenever I open NEI:

[NEI Subset Item Allocation/DEBUG]: Playing sound minecraft:sounds/random/break.ogg for event minecraft:random.break as channel cdf834df-ffda-40ae-92ac-d2d62e6fb1f7
[NEI Subset Item Allocation/DEBUG]: Playing sound minecraft:sounds/random/anvil_break.ogg for event minecraft:random.anvil_break as channel ee76e776-a6a3-49d3-93fd-ab2039ddbabf
[NEI Subset Item Allocation/DEBUG]: Playing sound ic2:sounds/Treetap.ogg for event ic2:tools.Treetap as channel 8013a018-9977-4f77-bf04-3f78922b6e26

Here's hoping RWTema reads this :)

**Comment by Dream-Master** [comment]
Maybe he read it but he doesn't do anything against it.

**Comment by KageDragon** [comment]
OK, did some digging in eclipse. 

[wall of text ahead!]

Here's whats happening, as far as I can tell:

The ExU Extra NEI Filtering seems to be a byproduct of the Advanced Item Filter code.
One of the options is HasContainerItem, which is any item that gets returned from a crafting operation
ie empty bucket when a water bucket is used in a recipe, or when a tool is returned, but with some durability used. 

ExU calls hasContainerItem for the GregTech metatool

GT's hasContainerItem method:

```
public final boolean hasContainerItem(ItemStack aStack) {
        return getContainerItem(aStack) != null;
    }
```

GT's getContainerItem method:

```
public final ItemStack getContainerItem(ItemStack aStack) {
        if (!isItemStackUsable(aStack)) return null;
        aStack = UT.Stacks.copyAmount(1, aStack);
        IToolStats tStats = getToolStats(aStack);
        if (tStats == null) return null;
        doDamage(aStack, tStats.getToolDamagePerContainerCraft());
        aStack = aStack.stackSize > 0 ? aStack : null;
        if (aStack == null) UT.Sounds.play(tStats.getBreakingSound(), 1, 1.0F); else UT.Sounds.play(tStats.getCraftingSound(), 1, 1.0F);
        return aStack;
    }
```

The last line is the relevant one, for tools, it plays break.ogg if the tool is out of uses after being used in a crafting operation.
getContainerItem plays a sound regardless if hasContainerItem is simply checking or if getContainerItem is being called from an actual crafting operation.

PrefixItem also has this issue, I think these are the glass tubes, and they're causing the Treetap.ogg to be played.

DarkAsp's GT5 version has a similar problem, and it was a simple fix:

https://github.com/Blood-Asp/GT5-Unofficial/commit/37def8854d1b3b33bc3f497d25a1f549c8240e04

There are two other issues however, and I think they're related:

The ExU extra filter categories I would assume would be calculated when a world is first loaded
and I assume should be static after that, yet every time NEI is opened, or I view a recipe\usage in NEI,
I get the sounds, which tells me getContainerItem is being called every time, I'm not sure exactly what ExU is doing...

I'll ask Greg and Tema what they think, has anyone had any luck contacting Tema through Twitter or Curse?

**Comment by plaguewolf** [comment]
nope. tema has left exU behind to pursue RL, or at least it appears so. might i know where you got the relevant source from? is there a developer build available? because id love to fix this myself, and just us it for myself of course. this is beyond annoying, and i can no longer simply use an outdated version of exU because it uses an old RFapi, and enderIO complains something fierce about it.

**Comment by plaguewolf** [comment]
nevermind, ive found the source, now im purusing the source, hopefully i can figure this out, and in addition maybe figure out why the golden bag with inventory tweaks causes, when inserted into itself, a divide by zero effect.

**Comment by Morketh** [comment]
Im fairly certain a lot of people would be happy for a patch if you get one put together.

**Comment by plaguewolf** [comment]
well that was funky, i wonder if its related...?

dev environ: buildpath: exU, GTUnofficial 5.09.09, mod path: NEI, ic2-exp, BC, forestry

gonna trim some mods off

java.util.ConcurrentModificationException
    at com.google.common.collect.HashBiMap$Itr.checkForConcurrentModification(HashBiMap.java:339) ~[guava-17.0.jar:?]
    at com.google.common.collect.HashBiMap$Itr.hasNext(HashBiMap.java:345) ~[guava-17.0.jar:?]
    at net.minecraft.client.audio.SoundManager.updateAllSounds(SoundManager.java:239) ~[SoundManager.class:?]
    at net.minecraft.client.audio.SoundHandler.update(SoundHandler.java:224) ~[SoundHandler.class:?]
    at net.minecraft.client.Minecraft.runTick(Minecraft.java:2104) ~[Minecraft.class:?]
    at net.minecraft.client.Minecraft.runGameLoop(Minecraft.java:1039) ~[Minecraft.class:?]
    at net.minecraft.client.Minecraft.run(Minecraft.java:962) [Minecraft.class:?]
    at net.minecraft.client.main.Main.main(Main.java:164) [Main.class:?]
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_60]
    at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_60]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_60]
    at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_60]
    at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
    at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
    at net.minecraftforge.gradle.GradleStartCommon.launch(Unknown Source) [start/:?]
    at GradleStart.main(Unknown Source) [start/:?]

**Comment by plaguewolf** [comment]
@KageDragon, what relevant class did you get that from? so far i've only found reference to hasContainerItem inside cofh.api.\* and inside ItemDivisionSigil.class

**Comment by plaguewolf** [comment]
com.rwtema.extrautils.tileentity.transfernodes.TileEntityTransferNodeInventory.class
oh dear lord, wall of if..else statements @ doCraft() has relevant hasContainerItem bits...

**Comment by KageDragon** [comment]
com\rwtema\extrautils\item\filters\AdvancedNodeUpgrades.class

```
addEntry(new Matcher("HasContainerItem")
{
  public boolean matchItem(ItemStack item)
  {
    return item.getItem().hasContainerItem(item);
  }
});
```

This was where whatever played the sound in GT was being called from, and I was focusing on the code that changed from 1.2.6->1.2.7

**Comment by plaguewolf** [comment]
ahhh, i see, i hadnt thought to compare the 2, ive just been looking through the code for 1.2.11 (or whatever the latest is)

**Comment by plaguewolf** [comment]
sadly, i realized earlier that lately i spend more time trying to fix my modpack, and reading through code, than actually playing minecraft. le sigh.

**Comment by ultrasn0w** [comment]
~~http://forum.industrial-craft.net/index.php?page=Thread&postID=195566#post195566
@rwtema, If you do not approve a patched extrautilites version, one post/comment will be enough and it will be gone again.
Oh and do NOT report bugs to rwtema when using this!!~~
Deleted patched version.

Fixed with version 1.2.12. Thank you @rwtema, the new feature is also working great!

**Comment by Morketh** [comment]
Nice patch and thank you for figuring that out

**Comment by ultrasn0w** [comment]
Confirmed that version 1.2.12 fixed this issue.

**Comment by SapientMC** [comment]
Same issue just after initialization of ME-system. Removing ME-controllers = quite, placing them back - annoying sound after few seconds. I've tried every version from 1.2.5 to 1.2.12 and nothing changes.

1-st time it appeared was crafting GT Motor via ME.

**Comment by JasonMcRay** [comment]
After some testing it is fixed in ExU 1.2.12 and newest GT
